core/src/trezor/wire/thp/pairing_context.py

```diff
@@ -0,0 +1,262 @@
+from typing import TYPE_CHECKING
+from ubinascii import hexlify
+
+import trezorui2
+from trezor import loop, protobuf, workflow
+from trezor.crypto import random
+from trezor.wire import context, message_handler, protocol_common
+from trezor.wire.context import UnexpectedMessageException
+from trezor.wire.errors import ActionCancelled, SilentError
+from trezor.wire.protocol_common import Context, Message
+
+if TYPE_CHECKING:
+    from typing import Container
+
+    from trezor import ui
+
+    from .channel import Channel
+    from .cpace import Cpace
+
+    pass
+
+if __debug__:
+    from trezor import log
+
+
+class PairingDisplayData:
+
+    def __init__(self) -> None:
+        self.code_code_entry: int | None = None
+        self.code_qr_code: bytes | None = None
+        self.code_nfc_unidirectional: bytes | None = None
+
+    def get_display_layout(self) -> ui.Layout:
+        from trezor import ui
+
+        # TODO have different layouts when there is only QR code or only Code Entry
+        qr_str = ""
+        code_str = ""
+        if self.code_qr_code is not None:
+            qr_str = self._get_code_qr_code_str()
+        if self.code_code_entry is not None:
+            code_str = self._get_code_code_entry_str()
+
+        return ui.Layout(
+            trezorui2.show_address_details(  # noqa
+                qr_title="Scan QR code to pair",
+                address=qr_str,
+                case_sensitive=True,
+                details_title="",
+                account="Code to rewrite:\n" + code_str,
+                path="",
+                xpubs=[],
+            )
+        )
+
+    def _get_code_code_entry_str(self) -> str:
+        if self.code_code_entry is not None:
+            code_str = f"{self.code_code_entry:06}"
+            if __debug__:
+                log.debug(__name__, "code_code_entry: %s", code_str)
+
+            return code_str[:3] + " " + code_str[3:]
+        raise Exception("Code entry string is not available")
+
+    def _get_code_qr_code_str(self) -> str:
+        if self.code_qr_code is not None:
+            code_str = (hexlify(self.code_qr_code)).decode("utf-8")
+            if __debug__:
+                log.debug(__name__, "code_qr_code_hexlified: %s", code_str)
+            return code_str
+        raise Exception("QR code string is not available")
+
+
+class PairingContext(Context):
+
+    def __init__(self, channel_ctx: Channel) -> None:
+        super().__init__(channel_ctx.iface, channel_ctx.channel_id)
+        self.channel_ctx: Channel = channel_ctx
+        self.incoming_message = loop.mailbox()
+        self.secret: bytes = random.bytes(16)
+
+        self.display_data: PairingDisplayData = PairingDisplayData()
+        self.cpace: Cpace
+        self.host_name: str
+
+    async def handle(self, is_debug_session: bool = False) -> None:
```

Comment:
File looks similar to code in `message_handler.py`. Might help with refactoring, we'll probably need something similar for BLE management context anyway.